src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,171 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::fabric::Network,
+    protocol::{
+        context::ProtocolContext,
+        reveal::reveal,
+        sort::ApplyInvStep::{RevealPermutation, ShuffleInputs, ShufflePermutation},
+        RecordId,
+    },
+    secret_sharing::Replicated,
+};
+use embed_doc_image::embed_doc_image;
+use futures::future::try_join_all;
+use permutation::Permutation;
+
+use super::{
+    apply::apply,
+    shuffle::{generate_random_permutation, Shuffle},
+};
+use tokio::try_join;
+
+/// This is APPLYINV(Algorithm 4) described in <https://eprint.iacr.org/2019/695.pdf>.
+/// This is a protocol that applies the inverse of a shared permutation to a shared-vector representation.
+/// Input: Each helpers know their own secret shares of input and permutation
+/// Output: At the end of the protocol, all helpers receive inputs after the permutation is applied
+#[derive(Debug)]
+#[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+pub struct SecureApplyInv<'a, F> {
+    input: &'a mut Vec<Replicated<F>>,
+    permutation: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> SecureApplyInv<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a mut Vec<Replicated<F>>, permutation: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { input, permutation }
+    }
+    /// This algorithm applies a permutation on the inputs. The permutation is not available in clear and
+    /// we do not want to reveal the actual permutation to any of the helpers while applying it.
```

Comment:
/// 
/// This algorithm applies a permutation to the `input` vector. The permutation is secret-shared, 
/// and none of the helpers should learn it through this protocol.
/// To keep the permutation secret, it (and the inputs) are first randomly securely shuffled.
/// After this shuffle, the permutation can be revealed. 
/// An adversary can only obtain a shuffled permutation, which is just a random permutation.

Reply:
Thanks for a better comment message!